packit/schema.py

```diff
@@ -483,18 +486,33 @@ class PackageConfigSchema(CommonConfigSchema):
     """
 
     jobs = fields.Nested(JobConfigSchema, many=True)
+    packages = fields.Dict(
+        keys=fields.String(), values=fields.Nested(CommonConfigSchema())
+    )
 
     # list of deprecated keys and their replacement (new,old)
     deprecated_keys = (("upstream_package_name", "upstream_project_name"),)
 
     @pre_load
-    def ordered_preprocess(self, data, **kwargs):
-        data = self.rename_deprecated_keys(data, **kwargs)
+    def ordered_preprocess(self, data: dict, **_) -> dict:
+        data = self.rename_deprecated_keys(data)
+        # Don't use 'setdefault' in this case, as we should expect
+        # downstream_package_name only if there is no 'packages' key.
+        if "packages" not in data:
+            package_name = data.pop("downstream_package_name")
+            data["packages"] = {
+                package_name: {
+                    "downstream_package_name": package_name,
+                }
+            }
         data.setdefault("jobs", get_default_jobs())
-        data = self.propagate_options_to_jobs(data, **kwargs)
+        # By this point, we expect both 'packages' and 'jobs' to be present
+        # in the config.
+        data = self.rearrange_packages(data)
+        data = self.rearrange_jobs(data)
         return data
```

Comment:
just to be clear, is it expected to pass `data` here, process both "packages" and "jobs" inside the "rearrange" methods and return the final `data` dict?

I am mostly curious about the `.pop` calls in both rearrange methods.

Reply:
Yes, that is expected.

Although strictly speaking `rearrange_packages` doesn't touch or need anything from `jobs`. It just pops it out (together with `packages`) so that `data` can be used in the merge with all the other top-level keys left. Probably this could have been implemented in some other ways too, but I found using this "pop and dict-merge" combination more elegant than a "loop and if" approach.

`rearrange_jobs` needs the content of the `packages` key in order to copy those to the job-objects. Although in that case pop-ing the keys was really not necessary, references work perfectly well. Thanks for pointing this out!